### ISSUE TITLE
Using latest Immutable with Typescript 2.4.1 causes Typescript to lock up/fail

### ISSUE THREAD

**Issue by AngryLawyer** [facebook/immutable-js#1265]
### What happened
Switching to Typescript 2.4.1 causes all ts tests to lock up/fail. It appears to be causing node to run out of memory.
This can also be reproduced by importing Immutable into a project using Typescript 2.4.1, in which case the reported error is:
```
<--- Last few GCs --->

  130691 ms: Mark-sweep 1317.5 (1434.2) -> 1317.5 (1434.2) MB, 2101.4 / 0 ms [allocation failure] [GC in old space requested].
  132883 ms: Mark-sweep 1317.5 (1434.2) -> 1317.5 (1434.2) MB, 2191.1 / 0 ms [allocation failure] [GC in old space requested].
  135075 ms: Mark-sweep 1317.5 (1434.2) -> 1317.5 (1434.2) MB, 2192.7 / 0 ms [last resort gc].
  137266 ms: Mark-sweep 1317.5 (1434.2) -> 1317.5 (1434.2) MB, 2190.7 / 0 ms [last resort gc].


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x2b1da96c9e59 <JS Object>
    1: instantiateTypeNoAlias [/Users/tony/Projects/x/manage/node_modules/typescript/lib/typescript.js:~34242] [pc=0x3abaf24cf038] (this=0x2b1da96e4dc9 <JS Global Object>,type=0xd201c537129 <a TypeObject with map 0x2422a10a4c41>,mapper=0x1769d2598f09 <JS Function (SharedFunctionInfo 0x356acb1605b1)>)
    2: instantiateType [/Users/tony/Projects/x/manage/node_modules/type...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
sh: line 1: 16214 Abort trap: 6           npm run lint-js
```

### How to reproduce
Edit the package.json, switch the Typescript version to 2.4.1.

**Comment by sccolbert** [comment]
c.f. https://github.com/Microsoft/TypeScript/issues/17070

**Comment by leebyron** [comment]
Yikes. That's not good. I'm not sure what the issue is, but very open to suggestions or PRs to fix this. 

Does anyone know what changed in this point release?

**Comment by ragtime** [comment]
Try to disable the (new) StrictGenericChecks by adding `"noStrictGenericChecks": true` to your `tsconfig.json`.

see https://github.com/Microsoft/TypeScript/issues/16777

**Comment by goloveychuk** [comment]
@leebyron they improved generics checking.
Looks this problem is because of cycle references and generics in it.
Should point, that it's not a inifinite cycle. If I reduce some code in immutable.d.ts it finish checking process (in a long time).

**Comment by JLarky** [comment]
Only information I can see for workaround is https://github.com/Microsoft/TypeScript/issues/17097#issuecomment-316857733

**Comment by sandersn** [comment]
Typescript 2.5.3 will contain a fix for this when it ships: Microsoft/Typescript#17984, among others.

**Comment by JLarky** [comment]
clickable link: https://github.com/Microsoft/TypeScript/pull/17984

**Comment by leebyron** [comment]
Pretty wild!

This should be fixed by #1285 - but please report back and reopen if the master build (install from https://github.com/facebook/immutable-js/tree/npm) doesn't solve the issue
